### PR TITLE
[Test]Fix e2e test

### PR DESCRIPTION
**Pull request by rui5i**
This PR adds recursion sample test back to e2e test. To prevent infinite loop, adding max_cache_staleness=0 to recursion sample pipeline step to disable caching.

Previous change: https://github.com/kubeflow/pipelines/pull/3433/files#diff-baf44ad432ad2937d5979e731be15648L134

### CODE REVIEW

File: samples/core/recursion/recursion.py [hunk 1]
```diff
@@ -59,6 +59,8 @@ def flip_component(flip_result):
 )
 def flipcoin():
     first_flip = flip_coin_op()
+    # set max_cache_staleness to 0 to prevent infinite loop due to caching
+    first_flip.execution_options.caching_strategy.max_cache_staleness = "P0D"
```
Comment (Ark-kun) [inline samples/core/recursion/recursion.py:63]: I think you might also need to fix the `flipA` task.

**Comment by numerology** [comment]
I would recommend to add a comment on top of the file that caching is tricky when recursion is involved.

And perhaps add a warning [here](https://github.com/kubeflow/pipelines/blob/fc5f977b1968ebcf4ada4b502016ac49e3cc0e46/sdk/python/kfp/dsl/_component.py#L101) as well? Up to you.

**Comment by numerology** [comment]
Or perhaps a more sophisticated solution:

When constructing a graph component/task, all sub-tasks will be default to having max_cache_staleness=0. WDYT? @Ark-kun  This can be a done in a separate PR.

**Comment by Ark-kun** [comment]
/lgtm

I think that this problem is not an issue for any real pipeline. For it to occur, the components and pipeline must be very artificial: The component must only have small value inputs and outputs that are ~same every time or not pass any data between components. Recursive coin toss might have the problem (since the tosses are not even connected), but anything real-world will probably not have an issue.

Our general guidance can be as follows: "Component code should be pure and deterministic in the sense that it produces same outputs given same inputs. If your component is not deterministic (for example, it returns a different random number on every invocation) you might want to disable caching for the tasks created from this component by setting `some_task.execution_options.caching_strategy.max_cache_staleness = "P0D"`".

WDYT?

**Comment by Ark-kun** [comment]
> When constructing a graph component/task, all sub-tasks will be default to having max_cache_staleness=0. WDYT? @Ark-kun

I have thought about this, but I think that caching would be desirable for most recursive loop scenarios while problematic scenarios are very rare.

Let's see the feedback. If we get feedback about people having problems with recursive loop caching, we can consider making the changes.

**Comment by numerology** [comment]
> /lgtm
> 
> I think that this problem is not an issue for any real pipeline. For it to occur, the components and pipeline must be very artificial: The component must only have small value inputs and outputs that are ~same every time or not pass any data between components. Recursive coin toss might have the problem (since the tosses are not even connected), but anything real-world will probably not have an issue.
> 
> Our general guidance can be as follows: "Component code should be pure and deterministic in the sense that it produces same outputs given same inputs. If your component is not deterministic (for example, it returns a different random number on every invocation) you might want to disable caching for the tasks created from this component by setting `some_task.execution_options.caching_strategy.max_cache_staleness = "P0D"`".
> 
> WDYT?

I'm okay with this. Also, generally speaking, IMO if the component logic is intended to be random (coin toss/cross validation with random shuffle etc.), caching should be disabled.

**Pull request merged**